torchmetrics/regression/cosine_similarity.py

```diff
@@ -0,0 +1,106 @@
+
+import torch
+from torch import Tensor
+
+from torchmetrics.functional.regression.cosine_similarity import _cosine_similarity_compute, _cosine_similarity_update
+from torchmetrics.metric import Metric
+from torchmetrics.utilities.data import dim_zero_cat
+
+
+class CosineSimilarity(Metric):
+    r"""
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+
+        .. math::
+            cos_{sim}(x,y) = \frac{x \cdot y}{||x|| \cdot ||y|| = \frac{\sum_{i=1}^n x_i y_i}{\sqrt{\sum_{i=1}^n x_i^2}
+             \sqrt{\sum_{i=1}^n y_i^2}}
+
+        where :math:`y` is a tensor of target values, and :math:`x` is a tensor of predictions.
+
+        Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
+
+       Args:
+           reduction : how to reduce over the batch dimension using 'sum', 'mean' or 'none'
+                        (taking the individual scores)
+           compute_on_step:
+               Forward only calls ``update()`` and return ``None`` if this is set to ``False``.
+           dist_sync_on_step:
+               Synchronize metric state across processes at each ``forward()``
+               before returning the value at the step.
+           process_group:
+               Specify the process group on which synchronization is called.
+               default: ``None`` (which selects the entire world)
+           dist_sync_fn:
+               Callback that performs the allgather operation on the metric state. When ``None``, DDP
+               will be used to perform the all gather.
+
+       Example:
+           >>> from torchmetrics import CosineSimilarity
+           >>> target = torch.tensor([[0, 1], [1, 1]])
+           >>> preds = torch.tensor([[0, 1], [0, 1]])
+           >>> cosine_similarity = CosineSimilarity(reduction = 'mean')
+           >>> cosine_similarity(preds, target)
+           tensor(0.8536)
+    """
+
+    def __init__(
+        self,
+        reduction: str = 'sum',
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn
+        )
+
+        self.add_state("preds", [], dist_reduce_fx="cat")
+        self.add_state("target", [], dist_reduce_fx="cat")
+        self.reduction = reduction
+
+    def update(self, preds: Tensor, target: Tensor):
+        """
+        Update state with predictions and targets. Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
+
+        Args:
+            preds: Predictions from model (probabilities, logits or labels)
+            target: Ground truth labels
+        """
```

Comment:
```suggestion
        """
        Update metric states with predictions and targets.
        Args:
            preds: Predicted tensor with shape ``(N,d)``
            target: Ground truth tensor with shape ``(N,d)``
        """
```

Reply:
@SkafteNicki alright, working on it! Is there anything more needed to be done? for example I saw a failed check on tests with ddp, which I could not run locally as I'm working on a Windows machine